CBMC: Add proof and contract for polyveck_invntt_tomont

jakemas opened the pull request:
Resolves https://github.com/pq-code-package/mldsa-native/issues/130

hanno-becker reviewed (commented):
@jakemas Did you try just unrolling the loop?

jakemas commented:
> @jakemas Did you try just unrolling the loop?

I'm following precedent here with other proofs in the directory for consistency and clarity. Not unrolling is consistant with:
- [polyvecl_ntt](https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/polyvec.c#L100-L117)
- [polyveck_ntt](https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/polyvec.c#L279-L296)
- [polyvecl_reduce](https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/polyvec.c#L63-L81)
- [polyveck_reduce](https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/polyvec.c#L194-L213)
- [polyvecl_invntt_tomont](https://github.com/pq-code-package/mldsa-native/blob/main/mldsa/polyvec.c#L119-L136)

hanno-becker commented:
@jakemas Sure, but did you try it?

jakemas commented:
> @jakemas Sure, but did you try it? It may well be that not unrolling is best, but I want to make sure we considered the alternative, esp. given the complex loop invariant.

Yes, I was able to get the proof to verify by unrolling the loop:
```
void polyveck_caddq(polyveck *v)
{

  if (MLDSA_K > 0) poly_invntt_tomont(&v->vec[0]);
  if (MLDSA_K > 1) poly_invntt_tomont(&v->vec[1]);
  if (MLDSA_K > 2) poly_invntt_tomont(&v->vec[2]);
  if (MLDSA_K > 3) poly_invntt_tomont(&v->vec[3]);
  if (MLDSA_K > 4) poly_invntt_tomont(&v->vec[4]);
  if (MLDSA_K > 5) poly_invntt_tomont(&v->vec[5]);
  if (MLDSA_K > 6) poly_invntt_tomont(&v->vec[6]);

}
```

hanno-becker commented:
@jakemas I'm not saying your current version is unsuitable, and we may well merge it as-is. I'm merely curious about alternative approaches to compare performance, esp. seeing that we currently resort to a copy to avoid a performance issue. In my experience, using CBMC involves a lot of trial and error and experimentation.

By unrolling, I did not mean unrolling in C, but in CBMC -- you can pass a flag to CBMC to ask it to unroll under the hood. You will find examples for this by grepping for `UNWINDSET` in the mlkem-native proof files.

jakemas commented:
> @jakemas I'm not saying your current version is unsuitable, and we may well merge it as-is. I'm merely curious about alternative approaches to compare performance, esp. seeing that we currently resort to a copy to avoid a performance issue. In my experience, using CBMC involves a lot of trial and error and experimentation.
> 
> By unrolling, I did not mean unrolling in C, but in CBMC -- you can pass a flag to CBMC to ask it to unroll under the hood; the invariant is then unneeded/unused. You will find examples for this by grepping for `UNWINDSET` in the mlkem-native proof `Makefile`s.

I added: `UNWINDSET += $(MLD_NAMESPACE)polyveck_invntt_tomont.0:8 # Largest value of MLDSA_K` to the makefile and removed the loop invariant, but my proof ran indefinitely.

mkannwischer commented:
I also tried out UNWINDSET for this proof and here is what I got: 
```
$ tests cbmc --single-step -p polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-44   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-44   (native):            SUCCESS (after 12s)
INFO  > CBMC (1/1)         ML-DSA-65   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-65   (native):            SUCCESS (after 93s)
INFO  > CBMC (1/1)         ML-DSA-87   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-87   (native):            SUCCESS (after 490s)
All good
```

I think this takes a little too long compared to @jakemas' proof:
```
$ tests cbmc --single-step -p polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-44   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-44   (native):            SUCCESS (after 7s)
INFO  > CBMC (1/1)         ML-DSA-65   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-65   (native):            SUCCESS (after 6s)
INFO  > CBMC (1/1)         ML-DSA-87   (native):         Starting CBMC proof for polyveck_ntt
INFO  > CBMC (1/1)         ML-DSA-87   (native):            SUCCESS (after 6s)
All good!
```

Let's merge as is and revisit once https://github.com/pq-code-package/mldsa-native/issues/102 is resolved. 

Overall, we should try to use UNWINDSET as much as possible so we don't have to clutter the code with invariants hindering readability. I usually try UNWINDSET first and if that is too slow I switch to invariants. For ML-KEM UNWINDSET always worked well, while for ML-DSA it seems it rarely gives acceptable performance.

mkannwischer reviewed (approved):
LGTM. I added it to https://github.com/pq-code-package/mldsa-native/issues/102.

hanno-becker commented:
Thanks @mkannwischer @jakemas for checking whether UNWINDSET is an option here.

> Overall, we should try to use UNWINDSET as much as possible so we don't have to clutter the code with invariants hindering readability. I usually try UNWINDSET first and if that is too slow I switch to invariants.

Agree.
